### DISCUSSION TITLE
`npm install` runs on 401 error during renovate execution using a private NPM feed

### DISCUSSION THREAD

**Discussion by lukaskolafa** [renovatebot/renovate#24654] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.61.0

### If you're self-hosting Renovate, select which platform you are using.

Azure DevOps (dev.azure.com)

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

When renovate is executed in repositories containing NPM, it should update both package.json and package-lock.json files.

### What you tried so far.

# Scenario 1:

the repository contains a `.npmrc` file with this content

```
registry=https://pkgs.dev.azure.com/***/npm/registry/
always-auth=true
```

The pipeline calls task before npm execution `npmAuthenticate@0`

`config.js` contains a section with a token

```
hostRules: [ {             
            matchHost: 'https://pkgs.dev.azure.com/$(organizationName)/',
            username: 'VssSessionToken',
            password: process.env.TOKEN }]
```     
Result:
- pull request is created
- package from the private feed is recognized and updated => hostRules and token are working
- package-lock.json is not updated => npm install task in the renovate execution step fails with 401 code

# Scenario 2:

the repository contains a `.npmrc` file with this content

```
registry=https://pkgs.dev.azure.com/***/npm/registry/
always-auth=true

//pkgs.dev.azure.com/***/npm/registry/:_authToken=***

```
no npmAuthenticate@0 task in the pipeline

`config.js` contains a section with a token

```
hostRules: [ {             
            matchHost: 'https://pkgs.dev.azure.com/$(organizationName)/',
            username: 'VssSessionToken',
            password: process.env.TOKEN }]
```     

Result:
- pull request is created
- package from the private feed is recognized and updated => hostRules and token are working
- package-lock.json is updated => nice, but this is not how it should work. We cannot have the file `.npmrc` in the repository with a valid token. This scenario was executed to find the root cause only.

# Scenario 3:

as described in the documentation here: https://docs.renovatebot.com/modules/platform/azure/

the repository contains a `.npmrc` file with this content

```
registry=https://pkgs.dev.azure.com/***/npm/registry/
always-auth=true
```

The pipeline calls task before npm execution `npmAuthenticate@0`

`config.js` contains a section with a token

```
hostRules: [ {             
            matchHost: 'https://pkgs.dev.azure.com/$(organizationName)/',
            username: 'VssSessionToken',
            password: process.env.TOKEN }]
``` 

renovate is executed using the `npx --userconfig .npmrc renovate` command

Result:
- pull request is created
- package from the private feed is recognized and updated => hostRules and token are working
- package-lock.json is not updated => npm install task in the renovate execution step fails with 401 code

# Scenario 4:

the repository does NOT contain a `.npmrc` file 

the file .npmrc is created on the fly in the pipeline with this content

```
registry=https://pkgs.dev.azure.com/***/npm/registry/
always-auth=true
```

The pipeline calls task before npm execution `npmAuthenticate@0`

`config.js` contains a section with a token

```
hostRules: [ {             
            matchHost: 'https://pkgs.dev.azure.com/$(organizationName)/',
            username: 'VssSessionToken',
            password: process.env.TOKEN }]
``` 

Result:
- NO pull request is created => the .npmrc file with the link to the private feed is ignored by renovate

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Executing command (repository=Renovate Demo/npm7, branch=renovate/npm)
       "command": "npm install --package-lock-only --no-audit --ignore-scripts express@4.18.2 pg@8.11.3"
DEBUG: rawExec err (repository=Renovate Demo/npm7, branch=renovate/npm)
       "err": {
         "cmd": "/bin/sh -c npm install --package-lock-only --no-audit --ignore-scripts express@4.18.2 pg@8.11.3",
         "stderr": "npm ERR! code E401\nnpm ERR! Unable to authenticate, your authentication token seems to be invalid.\nnpm ERR! To correct this please trying logging in again with:\nnpm ERR!     npm login\n\nnpm ERR! A complete log of this run can be found in: /tmp/renovate/cache/others/npm/_logs/2023-09-22T08_41_24_412Z-debug-0.log\n",
         "stdout": "",
         "options": {
           "cwd": "/tmp/renovate/repos/azure/Renovate Demo/npm7",
           "encoding": "utf-8",
           "env": {
             "NPM_CONFIG_CACHE": "/tmp/renovate/cache/others/npm",
             "npm_config_store": "/tmp/renovate/cache/others/pnpm",
             "HOME": "/home/vsts",
             "PATH": "/home/vsts/work/1/s/.renovate/node_modules/.bin:/home/vsts/work/1/s/node_modules/.bin:/home/vsts/work/1/node_modules/.bin:/home/vsts/work/node_modules/.bin:/home/vsts/node_modules/.bin:/home/node_modules/.bin:/node_modules/.bin:/opt/hostedtoolcache/node/18.17.1/x64/lib/node_modules/npm/node_modules/@npmcli/run-script/lib/node-gyp-bin:/opt/hostedtoolcache/node/18.17.1/x64/bin:/home/vsts/.local/bin:/opt/pipx_bin:/home/vsts/.cargo/bin:/home/vsts/.config/composer/vendor/bin:/usr/local/.ghcup/bin:/home/vsts/.dotnet/tools:/snap/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin",
             "LANG": "C.UTF-8",
             "CONTAINERBASE_CACHE_DIR": "/tmp/renovate/cache/containerbase"
           },
           "maxBuffer": 10485760,
           "timeout": 900000
         },
         "exitCode": 1,
         "name": "ExecError",
         "message": "Command failed: npm install --package-lock-only --no-audit --ignore-scripts express@4.18.2 pg@8.11.3\nnpm ERR! code E401\nnpm ERR! Unable to authenticate, your authentication token seems to be invalid.\nnpm ERR! To correct this please trying logging in again with:\nnpm ERR!     npm login\n\nnpm ERR! A complete log of this run can be found in: /tmp/renovate/cache/others/npm/_logs/2023-09-22T08_41_24_412Z-debug-0.log\n",
         "stack": "ExecError: Command failed: npm install --package-lock-only --no-audit --ignore-scripts express@4.18.2 pg@8.11.3\nnpm ERR! code E401\nnpm ERR! Unable to authenticate, your authentication token seems to be invalid.\nnpm ERR! To correct this please trying logging in again with:\nnpm ERR!     npm login\n\nnpm ERR! A complete log of this run can be found in: /tmp/renovate/cache/others/npm/_logs/2023-09-22T08_41_24_412Z-debug-0.log\n\n    at ChildProcess.<anonymous> (/home/vsts/work/1/s/.renovate/node_modules/renovate/lib/util/exec/common.ts:99:11)\n    at ChildProcess.emit (node:events:526:35)\n    at ChildProcess.emit (node:domain:489:12)\n    at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)"
       },

```

</details>

**Comment by viceice** [comment]
matchHost value from hostrule must be equal to the registry url from npmrc 

https://docs.renovatebot.com/getting-started/private-packages/#npm

**Reply by lukaskolafa** [reply]
Thank you, Michael, for the quick and helpful response. Indeed it worked.

Is the requirement to match the URL exactly between .npmrc and hostRules given by Renovate or by the executed npm install task? If Renovate gives the limitation, what is the reason? Is that something that I could change in the Renovate code?

The reason why I see this requirement as limiting is the following:
- it is very practical to define the host rules in Azure DevOps with this URL: `https://pkgs.dev.azure.com/$(organizationName)/` - the reason is that the part behind organizationName is dynamic, and if given exactly, it would not match some technologies - e.g., in case of NuGet feed, the middle part of the URL is different between package URL and NuGet feed URL. So using this common URL prefix makes it work for all cases.
- Conversely, the URL in `.npmrc` has to be a full package feed URL; otherwise, NPM does not find the packages.

in any case, I suggest updating this documentation: https://docs.renovatebot.com/modules/platform/azure/ as it states explicitly the URL has to match for yarn but does not mention npm, which confused me.

**Reply by rarkins** [reply]
It's a Renovate limitation, and a PR to loosen it would be welcome

**Reply by lukaskolafa** [reply]
I tried to find the correct place in the code, and it looks like it could be this place (please correct me if I am wrong):
- `lib\modules\manager\npm\post-update\rules.ts` [line 25]   (and the whole function `processHostRules` around it)

to me, it looks like it is actually an NPM restriction, not Renovate restriction. From what I see here, it looks like we just grab whatever is saved in the hostRules and use it as additional content for the final `.npmrc` file. And because I configure in the hostRules only partially matching URL (which is fine for Renovate), the `npm install` command does not recognize it and fails to authenticate.

so, to fix that, we would need to implement logic that grabs all hostRules and the existing `.npmrc` content, and both inputs would be combined into a working result. This could be done here `lib\modules\manager\npm\post-update\index.ts` [line 342]. To me, this looks like exactly what `npmAuthenticate@0` pipeline task would be doing, am I correct?

So what do you think, what is the correct strategy here? Should we try to somehow involve the standard pipeline task `npmAuthenticate@0` in the process or should we try to improve the .npmrc generation to support partially matching URLs? (of course, the has to apply to yarn, pnmp and others)

btw., meanwhile, I was able to get it running with a little workaround - using `npmrc` and `npmrcMerge` values in the `config.js`, which works fine for now, but has some other drawbacks..

**Reply by rarkins** [reply]
Here's what I think you'd need to do:

First, resolve the default registry (e.g. would be registry.npmjs.org if unspecified).
Query hostRules for that host, e.g. matchHost=https://registry.npmjs.org and hostType=npm.
If auth is returned, add it as default auth for the `.npmrc` file.

Next, iterate through the `.npmrc` for any non-default hosts and repeat the above hostRules steps for them.

The above algorithm should mean that your hostRules don't need to match the `.npmrc` exactly anymore
